OsmAnd/src/net/osmand/plus/plugins/development/widget/ZoomLevelWidget.java

```diff
@@ -15,52 +27,148 @@
 
 public class ZoomLevelWidget extends SimpleWidget {
 
+	private static final int ZOOM_OFFSET_FROM_31 = 17;
+	private static final int MAX_RATIO_DIGITS = 4;
+
 	private final OsmandMap osmandMap;
 	private final OsmandMapTileView mapView;
 
+	private final ZoomLevelWidgetState widgetState;
+
+	@Nullable
+	private ZoomLevelType cachedZoomLevelType;
 	private int cachedBaseZoom;
 	private int cachedZoom;
 	private float cachedZoomFloatPart;
 	private float cachedMapDensity;
+	private int cachedCenterX = -1;
+	private int cachedCenterY = -1;
 
-	public ZoomLevelWidget(@NonNull MapActivity mapActivity, @Nullable String customId, @Nullable WidgetsPanel widgetsPanel) {
+	public ZoomLevelWidget(@NonNull MapActivity mapActivity, @NonNull ZoomLevelWidgetState widgetState,
+	                       @Nullable String customId, @Nullable WidgetsPanel widgetsPanel) {
 		super(mapActivity, DEV_ZOOM_LEVEL, customId, widgetsPanel);
 		this.osmandMap = app.getOsmandMap();
 		this.mapView = mapActivity.getMapView();
+		this.widgetState = widgetState;
 		updateInfo(null);
 		setIcons(DEV_ZOOM_LEVEL);
+		setOnClickListener(getOnClickListener());
+	}
+
+	@NonNull
+	public OsmandPreference<ZoomLevelType> getZoomLevelTypePref() {
+		return widgetState.getZoomLevelTypePref();
+	}
+
+	@Override
+	protected OnClickListener getOnClickListener() {
+		return v -> {
+			widgetState.changeToNextState();
+			updateInfo(null);
+			updateWidgetName();
+		};
+	}
+
+	@Nullable
+	@Override
+	protected String getWidgetName() {
+		return widgetState != null ? widgetState.getTitle() : null;
+	}
+
+	@Nullable
+	@Override
+	public ZoomLevelWidgetState getWidgetState() {
+		return widgetState;
 	}
 
 	@Override
 	protected void updateSimpleWidgetInfo(@Nullable DrawSettings drawSettings) {
+		RotatedTileBox tileBox = mapView.getRotatedTileBox();
+		int newCenterX = tileBox.getCenter31X() >> ZOOM_OFFSET_FROM_31;
+		int newCenterY = tileBox.getCenter31Y() >> ZOOM_OFFSET_FROM_31;
+		ZoomLevelType newZoomLevelType = widgetState.getZoomLevelType();
 		int baseZoom = mapView.getBaseZoom();
 		int newZoom = mapView.getZoom();
 		float newZoomFloatPart = mapView.getZoomFloatPart() + mapView.getZoomAnimation();
 		float newMapDensity = osmandMap.getMapDensity();
-		if (isUpdateNeeded()
+
+		boolean update = isUpdateNeeded()
+				|| newZoomLevelType != cachedZoomLevelType
 				|| baseZoom != cachedBaseZoom
 				|| newZoom != cachedZoom
 				|| newZoomFloatPart != cachedZoomFloatPart
-				|| newMapDensity != cachedMapDensity) {
+				|| newMapDensity != cachedMapDensity
+				|| (newZoomLevelType == ZoomLevelType.MAP_SCALE && newCenterX != cachedCenterX || newCenterY != cachedCenterY);
```

Comment:
Looks somewhat strange, probably you missed "( )" when comparing centerX/centerY